### PR TITLE
fix: not handling `export default value` when `exportReferencedTypes` is `false`

### PR DESCRIPTION
**Pull request by samualtnorman**
this fixes
```ts
const foo = "bar"
export default foo
```
being compiled to
```ts
// Generated by dts-bundle-generator v6.10.0

declare const foo = "bar";
foo;

export {};
```

which causes `Error: Compiled with errors` from `error TS1036: Statements are not allowed in ambient contexts.`

it now compiles to
```ts
// Generated by dts-bundle-generator v6.10.0

declare const foo = "bar";

export {
	foo as default,
};

export {};
```

### CODE REVIEW

**Comment by samualtnorman** [comment]
sorry mislicked and pushed to `master`

what do you wanna do about the failing tests?
do they need to be changed or should I comply to them?

**Comment by timocov** [comment]
> do they need to be changed or should I comply to them?

First of all I'd say you need to add a test case for your specific case and run it locally. Are you sure that the issue exists? Probably the repro isn't so simple? I just tried to run your case locally and it works fine.

I think we need to address this first and based on this we'll decide what to do next.

**Comment by timocov** [comment]
```bash
$ cat index.ts
const foo = "bar"
export default foo

$ ./node_modules/.bin/dts-bundle-generator index.ts
Compiling input files...
Processing index.ts
Writing index.ts -> index.d.ts
Checking generated files...
Done in 13.25s

$ cat index.d.ts
// Generated by dts-bundle-generator v6.10.0

declare const foo = "bar";
export default foo;

export {};
```

Is there anything am I missing?

**Comment by samualtnorman** [comment]
how bizarre
the bug only happens when passing `--export-referenced-types`
sorry for my incompetence
I'm new to contributing to other people's projects

**Comment by samualtnorman** [comment]
what's the best way to go about creating a test for this?
I wonder if there are other bugs like this that slip by because they're not being tested with this flag

**Comment by timocov** [comment]
@samualtnorman that's alright, no worries. Let me guide you through this issue to make a proper fix 🙂 

First of all as I said you need to create a test case for your issue. See https://github.com/timocov/dts-bundle-generator/tree/master/tests/e2e for more information about tests and how to create a test case. In your case, don't forget to disable `exportReferencedTypes` option in the config file. Also, you can take a look at [this commit](https://github.com/timocov/dts-bundle-generator/commit/94618056c2b521d9e9437027b599e3a920a05424) where I added a new test case as a reference (or at any another commit in e2e folder).

To run tests you can use command `npm run test`. If you want to run only your test case, you can add `--grep` option to it `npm run test -- --grep TEST-NAME` (note that `--` is there for the purpose of as npm is passing arguments to an actual command). `TEST-NAME` there is the folder name (so choose a good name for your case 😄).

Before committing/pushing it is worth to check that all other tests are passing, thus please make sure that `npm run test` is passing after all fixes.

As for the issue itself and the ways to solve it. According to the fails of your fix it seems that the behaviour has been changed a bit and _probably_ now we have 2 branches of the code to handle the same thing, but perhaps one of them isn't used any more, which is not good I think.

But as you said that this only happens when `exportReferencedTypes` is disabled, it seems that the fix should be somewhere [where this option is using](https://github.com/timocov/dts-bundle-generator/blob/93c0208f2daef1dd74af2bc7f754d78ddf05d590/src/bundle-generator.ts#L301-L333). I'd suggest to take a look at this code. This function `shouldStatementHasExportKeyword` is using to detect whether a statement of the code should have an `export` keyword or it should/might be omitted. It looks like in the case of checking for `export default` statements it works incorrect thus I'd advise you to check it.

If you'll have any questions, findings, ideas, or you decide to give up - just reply here and I'll do my best to help you to solve it.

**Comment by timocov** [comment]
I'd start from looking at the value of `statementExports` [here](https://github.com/timocov/dts-bundle-generator/blob/93c0208f2daef1dd74af2bc7f754d78ddf05d590/src/bundle-generator.ts#L301) for `export default` statement.

To debug the code you can do the following:
- Run `node --inspect-brk ./node_modules/mocha/bin/_mocha --grep TEST-NAME` (where `TEST-NAME` is the name of your new test)
- Open Chrome DevTools
- Click on this button to open dedicated DevTools for nodejs 
image
- Then you can debug the tool for your test case (e.g. set up breakpoints at the code I've sent above)

I hope this helps! Also, to better understand typescript's API and AST (and all other related internals) I'd suggest to use https://ts-ast-viewer.com/ as it provides you all the information for any node in the code so this is extremely helpful tool in this matter.

TIP: instead of setting up breakpoints in the DevTools I used to use a `debugger` statement right in the code since in this case you don't need to wait until nodejs loads all the code to set them up (but please don't commit it 😆)

**Comment by timocov** [comment]
Another TIP (if you aren't aware yet): every ast node in TypeScript has a method called `getText` that returns you a string for this node so it might be helpful while debugging to distinguish between nodes/statements in a quicker way.

**Comment by samualtnorman** [comment]
ugh I accidentally pressed the same button again
please ignore
I made a mistake doing this on `master`

**Comment by samualtnorman** [comment]
so I think the issue stems from stuff checking if the statement has a default modifier (whatever that is), and export assignments just don't have a default modifier so the check fails

**Comment by samualtnorman** [comment]
so something else that sucks is that export assignments don't have a `.name` for some reason

**Comment by samualtnorman** [comment]
tests are passing but I'm not sure my solution is correct

File: tests/e2e/test-cases/export-default-no-export-referenced-types/config.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { TestCaseConfig } from '../test-case-config';
+
+const config: TestCaseConfig = {
+	failOnClass: true,
```
Comment (timocov) [inline tests/e2e/test-cases/export-default-no-export-referenced-types/config.ts:4]: is this related to the test case?

File: src/helpers/typescript.ts [hunk 2]
```diff
@@ -27,6 +27,9 @@ export function hasNodeModifier(node: ts.Node, modifier: ts.SyntaxKind): boolean
 export function getNodeName(node: ts.Node): NodeName | undefined {
 	const nodeName = (node as unknown as ts.NamedDeclaration).name;
 	if (nodeName === undefined) {
+		if (ts.isExportAssignment(node) && !node.isExportEquals) {
+			return node.expression as ts.Identifier;
```
Comment (timocov) [inline src/helpers/typescript.ts:31]: how does it work without this fix? Since you're handling a special case for `export default name` statement and assuming that `getNodeName` returns a name of "exported" node I'd expect to return a `default` keyword or string from here (as it is at L35), so I'm just curious what exactly is broken without this change.

File: tests/e2e/test-cases/export-default-no-export-referenced-types/config.ts (hunk 1 above)
Comment (samualtnorman) [reply tests/e2e/test-cases/export-default-no-export-referenced-types/config.ts:4]: ah, I just copy pasted from the sample

File: src/helpers/typescript.ts (hunk 2 above)
Comment (samualtnorman) [reply src/helpers/typescript.ts:31]: so the issue is that `.find()` returns `undefined` and we fallthrough and return `undefined`
I initially thought `getNodeName()` was supposed to return the `default` token
but `exportedSymbols` is an array of 1 symbol, and that symbol is the `foo` one
so I knew I just needed to return that identifier otherwise the `.filter()` in `getExportsForName()` returns an empty array

Comment (samualtnorman) [reply src/helpers/typescript.ts:31]: in fact, `.modifiers` is `undefined` so `.find()` isn't even called

File: tests/e2e/test-cases/export-default-no-export-referenced-types/config.ts (hunk 1 above)
Comment (timocov) [reply tests/e2e/test-cases/export-default-no-export-referenced-types/config.ts:4]:
```suggestion
```

File: src/helpers/typescript.ts (hunk 2 above)
Comment (timocov) [reply src/helpers/typescript.ts:31]: yeah, fair enough, let me dive deeper what we need to do here. I'm thinking about rename probably (the function name) or handle it in a bit different way so we won't treat an expression of default export as a name and will handle it separately in the other place, where it is needed. I'll be back shorty

Comment (samualtnorman) [reply src/helpers/typescript.ts:31]: on the side, what's the point of
```ts
if (defaultModifier !== undefined) {
	return defaultModifier;
}
```

if we're `return`ing it anyway, surely we can just `return node.modifiers?.find((mod: ts.Modifier) => mod.kind === ts.SyntaxKind.DefaultKeyword);`?

Comment (timocov) [reply src/helpers/typescript.ts:31]: It seems that this should works fine without this change. The purpose of this function is to return a name of a statement that you can refer to in the code (or use to check what the name of the node is so you can compare it with the exported one). And for this kind of statements (`export default name`) the result is kind of UB because this node doesn't have a name and is kind of proxy to another node (or a symbol, whatever):

- For `export function foo ...` the result is the node for identifier `name` (because we're declaring a function in this first place with the name `foo` but additionally it is exported thus has a modifier `export`)
- For `export default asd;` the result cannot be determined because we cannot refer to a node that is `export default`, i.e. export assignment node in terms of the compiler's API. But you should be able to get a result for a declaration statement of `asd` node for sure (e.g. `class asd {}` or `function asd() {}` or `let asd` or `const asd` and so on) because you can refer to that nodes, they are kind of valuable in terms of declaration nodes/types if you wish (not sure how to explain it in a better way, sorry)

So in this case I'd even expect that this function should raise an error in the case when it has been called with this kind of nodes. But this is too much for this PR and this kind of change 😄 

As for a proper fix I think it would be better to do the following:
1. Revert the changed you made in these 2 files (`typescript.ts` and `bundle-generator.ts`) (I'm sorry)
1. Add additional checks to [this code](https://github.com/timocov/dts-bundle-generator/blob/93c0208f2daef1dd74af2bc7f754d78ddf05d590/src/bundle-generator.ts#L471-L495) and instead of handling both `export =` and `export default` similarly by adding their statements (from the root source file) to the result nodes, we could handle `export default` statement differently by resolving the original name and adding it as `export { name as default }`. This is similar to what you had in https://github.com/timocov/dts-bundle-generator/commit/e2045fc5e18c186c805255f538b30d6cac5289b7 (your original fix I guess?), but more "smarter" (to handle renames while importing).

And yes, it requires to update output of some of the test cases (instead of `export default name` it becomes `export { name as default }`, but 1) since the output is correct 2) looks good; I think this should be fine.

I'm going to push the changes to this PR (it looks the PR is from your master branch, are you ok with that?) since I already made them while checking this code but I'll put you as a co-author of the commit because I think it will be fair.

Comment (timocov) [reply src/helpers/typescript.ts:31]: @samualtnorman But nevertheless, I must say that you did a really good job and dove enough to understand and find _the correct_ roots of the problem. Even your first fix was too close to the solution I'd expect to have in the tool (and I'm really sorry for reverting your changes because of that, I do understand how is that to revert the work you've worked and invested some of your time). I really appreciate it ❤️!

Comment (samualtnorman) [reply src/helpers/typescript.ts:31]: nah I do not mind, thank you for looking into this with me

Comment (timocov) [reply src/helpers/typescript.ts]: Here we go ac8644d98d03eb6a8d6fc8d59e89ebdab7839440. So basically this code handles a situation when you import a node with a renaming and then export it as default:

```ts
import { originalName as newName } from './module';

export default newName;
```

In this case the correct output should be `export { originalName as default }` (since all the nodes are putting together without renaming) instead of `export { newName as default }`.

**Comment by timocov** [comment]
The fix has been deployed in [v6.11](https://github.com/timocov/dts-bundle-generator/releases/tag/v6.11.0). Many thanks for your contribution, @samualtnorman! ❤️
